### ISSUE TITLE
PyMC Student-T distribution does not have a JAX dispatch

### ISSUE THREAD

**Issue by HasnainRaz** [pymc-devs/pymc#7027]
### Before

```python
pm.sample_posterior_predictive(trace, random_seed=random_seed, compile_kwargs={"mode": "JAX"})

failing with models using the Student-T dist.
```

### After

```python
pm.sample_posterior_predictive(trace, random_seed=random_seed, compile_kwargs={"mode": "JAX"})

working with models using the Student-T dist.
```

### Context for the issue:

I have a regression model which requires me to sample from a student-t distribution.

When sampling the posterior predictive on a conditional version of this model (for forecasting), the sampling takes around 20 mins. Naturally, I would like to speed this up.

I looked and saw the sample_posterior_predictive in pymc supports "compile_kwargs" argument and I can use JAX to compile the graph (compile_kwargs={"mode": "JAX"}). However, when I do this, I get an error saying:
```
See traceback:

Traceback (most recent call last):
  File "/Users/hasnain.raza/Projects/test/train.py", line 213, in <module>
    fit(config, df, None)
  File "/Users/hasnain.raza/Projects/test/train.py", line 54, in fit
    trace, train_metrics = trainer.fit(
  File "/Users/hasnain.raza/Projects/test/trainer.py", line 45, in fit
    posterior_idata = pm.sample_posterior_predictive(trace, random_seed=random_seed, compile_kwargs={"mode": "JAX"})
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pymc/sampling/forward.py", line 644, in sample_posterior_predictive
    values = sampler_fn(**param)
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pymc/util.py", line 393, in wrapped
    return core_function(**input_point)
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/compile/function/types.py", line 970, in __call__
    self.vm()
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/link/utils.py", line 202, in streamline_default_f
    raise_with_op(fgraph, node, thunk)
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/link/utils.py", line 531, in raise_with_op
    raise exc_value.with_traceback(exc_trace)
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/link/utils.py", line 198, in streamline_default_f
    thunk()
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/link/basic.py", line 660, in thunk
    outputs = fgraph_jit(*[x[0] for x in thunk_inputs])
  File "/var/folders/2c/th0st_4j3hq6gjqljsvdly4r0000gq/T/tmptx0md65m", line 3, in jax_funcified_fgraph
    variable, y_pred = sample_fn(random_generator_shared_variable, tensor_constant, tensor_constant_1, tensor_constant_2, _, sigma)
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/link/jax/dispatch/random.py", line 111, in sample_fn
    return jax_sample_fn(op)(rng, out_size, out_dtype, *parameters)
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/functools.py", line 889, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/link/jax/dispatch/random.py", line 119, in jax_sample_fn
    raise NotImplementedError(
NotImplementedError: No JAX implementation for the given distribution: studentt
```

Similarly, using `compile_kwargs={"mode": "NUMBA"}` leads to:

```
Traceback (most recent call last):
  File "/Users/hasnain.raza/test/train.py", line 213, in <module>
    fit(config, df, None)
  File "/Users/hasnain.raza/Projects/test/train.py", line 54, in fit
    trace, train_metrics = trainer.fit(
  File "/Users/hasnain.raza/Projects/test/trainer.py", line 45, in fit
    posterior_idata = pm.sample_posterior_predictive(trace, random_seed=random_seed, compile_kwargs={"mode": "NUMBA"})
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pymc/sampling/forward.py", line 614, in sample_posterior_predictive
    _sampler_fn, volatile_basic_rvs = compile_forward_sampling_function(
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pymc/sampling/forward.py", line 263, in compile_forward_sampling_function
    compile_pymc(inputs, fg.outputs, givens=givens, on_unused_input="ignore", **kwargs),
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pymc/pytensorf.py", line 1195, in compile_pymc
    pytensor_function = pytensor.function(
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/compile/function/__init__.py", line 315, in function
    fn = pfunc(
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/compile/function/pfunc.py", line 469, in pfunc
    return orig_function(
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/compile/function/types.py", line 1756, in orig_function
    fn = m.create(defaults)
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/compile/function/types.py", line 1649, in create
    _fn, _i, _o = self.linker.make_thunk(
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/link/basic.py", line 245, in make_thunk
    return self.make_all(
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/link/basic.py", line 688, in make_all
    thunks, nodes, jit_fn = self.create_jitable_thunk(
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/link/basic.py", line 638, in create_jitable_thunk
    converted_fgraph = self.fgraph_convert(
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/link/numba/linker.py", line 27, in fgraph_convert
    return numba_funcify(fgraph, **kwargs)
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/functools.py", line 889, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/link/numba/dispatch/basic.py", line 459, in numba_funcify_FunctionGraph
    return fgraph_to_python(
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/link/utils.py", line 734, in fgraph_to_python
    compiled_func = op_conversion_fn(
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/functools.py", line 889, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/link/numba/dispatch/basic.py", line 428, in numba_funcify
    return generate_fallback_impl(op, node, storage_map, **kwargs)
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/link/numba/dispatch/basic.py", line 381, in generate_fallback_impl
    ret_sig = numba.types.Tuple([get_numba_type(o.type) for o in node.outputs])
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/link/numba/dispatch/basic.py", line 381, in <listcomp>
    ret_sig = numba.types.Tuple([get_numba_type(o.type) for o in node.outputs])
  File "/Users/hasnain.raza/anaconda3/envs/mmm/lib/python3.10/site-packages/pytensor/link/numba/dispatch/basic.py", line 119, in get_numba_type
    raise NotImplementedError(f"Numba type not implemented for {pytensor_type}")
NotImplementedError: Numba type not implemented for RandomGeneratorType
```

If I swap out the student-t and use a normal distribution, the JAX compilation works (numba still fails), and sampling finishes in 3 seconds. So you can see how this feature is really important to me, literally from 20 mins to 3 seconds. Sadly, the normal dist doesn't model my problem correctly, so I am stuck with the slow posterior predictive sampling.

Any suggestions here? Would it be possible to add a JAX dispatch for student-t?

**Comment by ricardoV94** [comment]
Should be fixed by #7028

**Comment by HasnainRaz** [comment]
@ricardoV94 Thanks for the reply.

I had the same reaction tbh, but I tested thoroughly (It's not the normal distribution making the difference, it IS the JAX compilation). Also no GPUs, I'm benchmarking on an M1.

Sadly I can't share the model for privacy reasons. I will try to come up with a minimal example to reproduce.

Also interesting that the problem is on pymc side... I assumed since it was raised in the pytensor module, it would be pytensor related. Thanks for triaging so quick

**Comment by ricardoV94** [comment]
No problem. It would be great to have a model (even if altered) that we could have a look at to understand the differences. Is it the compilation time that is taking a huge toll or the actual sampling?

**Comment by HasnainRaz** [comment]
@ricardoV94 It is the posterior predictive sampling (not the compilation), I can see the sampling progress bar taking 20 mins without JAX compilation

**Comment by ricardoV94** [comment]
Really surprising then. Do you get numpy-blas warnings when you import PyMC? My best guess is you have some heavy matrix operations and the compiled function is using slow python code to evaluate them. I can't imagine the difference in speed would come from the actual RVs in the graph.

**Comment by HasnainRaz** [comment]
@ricardoV94 Entirely possible.

Also I tried out that pull request #7028  and indeed it fixes the issue. Now I can sample in 3 seconds.
Though I am seeing that PyMC-experimental is broken by this change: https://github.com/pymc-devs/pymc/commit/04a03b520aa52c5586123f076c6377b32cf855b4

Because RVTransform is imported here: https://github.com/pymc-devs/pymc-experimental/blob/main/pymc_experimental/utils/prior.py#L22

I just fixed it locally to run it for testing, but thought I'd let you guys know too.
